![image.png](attachment:e63b5ccf-add2-4ee0-9d50-93748883cf5e.png)

Based on the formula for calculating transposed convolution:

$$
Output = (Input - 1) \times stride - 2 \times padding
 + dilation \times (kernel size - 1) + output padding + 1
$$

The input is 8 * 8, kernel size is 4, stride is 2, padding is 1 and output padding is 1, so the output size will be:

$(8-1) \times 2 - 2 \times 1 + 1 \times (4-1) + 1 + 1 = 17$

So the output size will be $17 \times 17$

![image.png](attachment:12fe1bec-990d-4323-9b27-025162c776f5.png)

If I increase the stride from 2 to 3, which means that the output will be more sparse, the output size will be:

(8-1) * 3 - 2 * 1 + 1 * (4-1) + 1 + 1 = 24

so the output size will be $24 \times 24$

![image.png](attachment:e06b3e8b-0abd-479b-ad83-e28b91174a23.png)

Formula will be:

$$
O = (I - 1) \times S - 2P + D \times (K - 1) + OP + 1
$$

If assuming that dilation d = 1, the formula will be:

$$
O = (I - 1) \times S - 2P + K + OP
$$

![image.png](attachment:713bdb8a-7ea8-494b-9b35-27656354a28b.png)

Based on the formula, assume that there's no padding, input size is $16 \times 16$, output size is $32 \times 32$

So one possible kernel size and stride will be (assuming output padding is also 0):

$32 = (16 - 1) \times S - 2 \times 0 + 1 \times (K - 1) + 0 + 1$

so we can have K = 2, S = 2

Be more detailed, kernel size = 2, stride = 2, padding = 0, output padding = 0

![image.png](attachment:554e8523-ac34-412a-a0cb-2b61ae073b64.png)

The formula for calculating the BatchNorm is:

$$
y_i = \gamma \frac{x_i - \mu_B}{\sqrt{\sigma_B^2 + \epsilon}} + \beta
$$

Since we are using zero-mean and unit-variance, meaning that we need this normalized batch to have mean = 0 and std = 1

Formula will be:
$$
y_i = \frac{x_i - \mu_B}{\sqrt{\sigma_B^2 + \epsilon}}
$$

assuming that $\epsilon$ is very small, we can ignore it, so the formula will be:
$$
y_i = \frac{x_i - \mu_B}{\sqrt{\sigma_B^2}}
$$

In [1]:
import numpy as np


def normalization(batch, gamma=1, beta=0, eps=1e-5):
    mean = np.mean(batch)
    var = np.var(batch)
    normalized_batch = (gamma * (batch - mean) / np.sqrt(var + eps)) + beta
    return normalized_batch

mini_batch = np.array([6,8,10,6])
norm = normalization(mini_batch, eps = 0) # normally don't set eps = 0, but here we can set it to 0 to make sure the normalized batch has mean = 0 and std = 1
print("normalized batch:", norm)

normalized batch: [-0.90453403  0.30151134  1.50755672 -0.90453403]


So the normalized output using BatchNorm is [-0.90453403, 0.30151134, 1.50755672, -0.90453403]

![image.png](attachment:8f16b63e-d7e0-4314-b531-8640afbcc88f.png)

The key mathematical difference between ReLU and LeakyReLU is that ReLU has a slope of 0 for negative inputs, while LeakyReLU has a small positive slope for negative inputs.

This means that LeakyReLU allows for a small amount of information to flow through the network even when the input is negative, which can help prevent the "dying ReLU" problem where neurons become inactive and stop learning.

Formula for ReLU:
$$
\text{ReLU}(x)=
\begin{cases}
x, & x>0 \\
0, & x\leq 0
\end{cases}
$$

Formula for LeakyReLU:
$$
\text{LeakyReLU}(x)=
\begin{cases}
x, & x>0 \\
\alpha x, & x\leq 0
\end{cases}
$$

where
$$
\alpha \in (0,1)
$$

![image.png](attachment:f1f157fb-13d4-4f74-9254-fac0772804f9.png)

LeakyReLU can help prevent the "dying ReLU" problem by allowing a small gradient to flow through the network even when the input is negative (LeakyReLU can aviods zero gradients for negative inputs). This can help keep neurons active and allow them to continue learning, which can improve the overall performance of the deep networks.